### PR TITLE
[Refactor upsamplers and downsamplers] separate out upsamplers and downsamplers.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Part of a series of refactoring PRs to be followed. This one just separates out the upsamplers and downsamplers to keep the `resnet.py` module clean and lean.

### CODE REVIEW

File: src/diffusers/models/resnet.py [hunk 1]
```diff
@@ -25,560 +25,17 @@
 from .attention_processor import SpatialNorm
 from .lora import LoRACompatibleConv, LoRACompatibleLinear
 from .normalization import AdaGroupNorm
-
```
Comment (sayakpaul) [inline src/diffusers/models/resnet.py:28]: 💯 percent not sure why this important change is now being displayed:

https://github.com/huggingface/diffusers/blob/2cac34f41929edb4041cdea9c67f3cfa7cc5dc66/src/diffusers/models/resnet.py#L28

**Review by DN6 — COMMENTED** [review]
Nice! I think the `upsample_2d` and `downsample_2d` functions can also be moved into the samples directory? 

They also only seem to be used with the fir kernel. So maybe we can rename them to be explicit?

File: src/diffusers/models/samplers/upsamplers.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        self.name = name
+        conv_cls = nn.Conv2d if USE_PEFT_BACKEND else LoRACompatibleConv
+
+        conv = None
+        if use_conv_transpose:
+            conv = nn.ConvTranspose2d(channels, self.out_channels, 4, 2, 1)
+        elif use_conv:
+            conv = conv_cls(self.channels, self.out_channels, 3, padding=1)
+
+        # TODO(Suraj, Patrick) - clean up after weight dicts are correctly renamed
```
Comment (yiyixuxu) [inline src/diffusers/models/samplers/upsamplers.py:114]: can we address this?

File: src/diffusers/models/samplers/upsamplers.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        if output_size is None:
+            hidden_states = F.interpolate(hidden_states, scale_factor=2.0, mode="nearest")
+        else:
+            hidden_states = F.interpolate(hidden_states, size=output_size, mode="nearest")
+
+        # If the input is bfloat16, we cast back to bfloat16
+        if dtype == torch.bfloat16:
+            hidden_states = hidden_states.to(dtype)
+
+        # TODO(Suraj, Patrick) - clean up after weight dicts are correctly renamed
```
Comment (yiyixuxu) [inline src/diffusers/models/samplers/upsamplers.py:153]: and this?

**Review by yiyixuxu — COMMENTED** [review]
thanks! I think there are still some to-do comments there needs to be addressed

**Comment by sayakpaul** [comment]
@yiyixuxu I don't have context for those TODOs actually. So, I'd prefer to address them separately. Cc: @patrickvonplaten @patil-suraj.

**Comment by sayakpaul** [comment]
> Nice! I think the `upsample_2d` and `downsample_2d` functions can also be moved into the samples directory?
> 
> They also only seem to be used with the fir kernel. So maybe we can rename them to be explicit?

@DN6 I have moved the `downsample_2d` and `upsample_2d` functions to `samplers/utils.py`. I think that's a good idea. However, I am not sure about explicitly renaming them. Even though they are not leveraged for FIR kernels, in the future, we might generalize them beyond just FIR kernels should an opportunity arise. So, I think it's better to leave them as is for now.

File: src/diffusers/models/samplers/utils.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Tuple
+
+import torch
+import torch.nn.functional as F
+
+
+def upfirdn2d_native(
```
Comment (patrickvonplaten) [inline src/diffusers/models/samplers/utils.py:21]: Let's put this also in the upsamplers file

File: src/diffusers/models/samplers/utils.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    out = out.permute(0, 2, 3, 1)
+    out = out[:, ::down_y, ::down_x, :]
+
+    out_h = (in_h * up_y + pad_y0 + pad_y1 - kernel_h) // down_y + 1
+    out_w = (in_w * up_x + pad_x0 + pad_x1 - kernel_w) // down_x + 1
+
+    return out.view(-1, channel, out_h, out_w)
+
+
+def upsample_2d(
```
Comment (patrickvonplaten) [inline src/diffusers/models/samplers/utils.py:71]: let's put this in the upsamplers file

File: src/diffusers/models/samplers/utils.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    output = upfirdn2d_native(
+        hidden_states,
+        kernel.to(device=hidden_states.device),
+        up=factor,
+        pad=((pad_value + 1) // 2 + factor - 1, pad_value // 2),
+    )
+    return output
+
+
+def downsample_2d(
```
Comment (patrickvonplaten) [inline src/diffusers/models/samplers/utils.py:118]: let's put this in the downsamplers file

File: src/diffusers/models/samplers/__init__.py [hunk 7]
```diff
@@ -0,0 +1,3 @@
+from .downsamplers import Downsample1D, Downsample2D, FirDownsample2D, KDownsample2D
```
Comment (patrickvonplaten) [inline src/diffusers/models/samplers/__init__.py:1]: I don't think "samplers" is a good name as many people understand a "sampler" as a "scheduler". Let's just keep the files on the src/diffusers/models level for the moment and just call them downsampling.py and upsampling.py

Comment (sayakpaul) [reply src/diffusers/models/samplers/__init__.py:1]: How about `interpolation`?

Comment (sayakpaul) [reply src/diffusers/models/samplers/__init__.py:1]: We could also do `blocks`? I think that's more generic. I'd really like to move these files to a separate module and keep `models` as lean as possible.

File: src/diffusers/models/samplers/utils.py (hunk 5 above)
Comment (sayakpaul) [reply src/diffusers/models/samplers/utils.py:71]: The thing is it's used in `downsamples.py::_downsample_2d` too.

File: src/diffusers/models/samplers/utils.py (hunk 6 above)
Comment (sayakpaul) [reply src/diffusers/models/samplers/utils.py]: Done in 3f961453.

File: src/diffusers/models/downsamplers.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (patrickvonplaten) [inline src/diffusers/models/downsamplers.py:1]: (nit) would prefer the filename `downsampling.py` here so that people 100% understand that we're not talking about "samplers"

Comment (sayakpaul) [reply src/diffusers/models/downsamplers.py]: 100 percent agreed here. Done in f200ce4c.

File: src/diffusers/models/upsampling.py [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+    out = out.permute(0, 2, 3, 1)
+    out = out[:, ::down_y, ::down_x, :]
+
+    out_h = (in_h * up_y + pad_y0 + pad_y1 - kernel_h) // down_y + 1
+    out_w = (in_w * up_x + pad_x0 + pad_x1 - kernel_w) // down_x + 1
+
+    return out.view(-1, channel, out_h, out_w)
+
+
+def upsample_2d(
```
Comment (DN6) [inline src/diffusers/models/upsampling.py:382]: I still think it would be better to be explicit and name this to something like `upsample_fir2d`. But not a blocker to merge.

**Review by DN6 — APPROVED** [review]
LGTM 👍🏽
